evm/p2p/cancel_token.py

```diff
@@ -0,0 +1,90 @@
+import asyncio
+from typing import Any, Awaitable
+
+from evm.p2p.exceptions import OperationCancelled
+
+
+class CancelToken:
+
+    def __init__(self, name: str, next_token: 'CancelToken' = None) -> None:
+        self.name = name
+        self.next = next_token
+        self._triggered = asyncio.Event()
+
+    def chain(self, next_token: 'CancelToken') -> 'CancelToken':
+        """Return a new CancelToken chaining this and next_token.
+
+        The new CancelToken's triggered will return True if trigger() has been
+        called on either of the chained tokens, but calling trigger() on this token
+        has no effect on either of the chained tokens.
+        """
+        return CancelToken(":".join([self.name, next_token.name]), next_token)
+
+    def trigger(self) -> None:
+        self._triggered.set()
+
+    @property
+    def triggered_token(self) -> 'CancelToken':
+        if self._triggered.is_set():
+            return self
+        elif self.next is not None:
+            return self.next.triggered_token
+        else:
+            return None
+
+    @property
+    def triggered(self) -> bool:
+        if self.next is None:
+            return self._triggered.is_set()
+        return self._triggered.is_set() or self.next.triggered
+
+    async def wait(self) -> None:
+        if self.triggered_token is not None:
+            return
+
+        futures = [asyncio.ensure_future(self._triggered.wait())]
+        if self.next is not None:
+            futures.append(asyncio.ensure_future(self.next.wait()))
+
+        def cancel_pending(f):
+            for future in futures:
+                if not future.done():
+                    future.cancel()
+
+        fut = asyncio.ensure_future(self._wait_for_first(futures))
+        fut.add_done_callback(cancel_pending)
+        await fut
+
+    async def _wait_for_first(self, futures):
+        for future in asyncio.as_completed(futures):
+            await future
+            return
+
+    def __str__(self):
+        return self.name
+
+
+async def wait_with_token(future: Awaitable,
+                          cancel_token: CancelToken,
+                          timeout: float = None) -> Any:
+    """Wait for future to complete, unless we timeout or the cancel token is triggered.
+
+    Returns the future's result in case it completes.
+
+    Raises TimeoutError if we timeout or OperationCancelled if the cancel token is triggered.
+    """
+    done, pending = await asyncio.wait(
+        [future, cancel_token.wait()],
+        timeout=timeout,
+        return_when=asyncio.FIRST_COMPLETED)
+    for task in pending:
+        task.cancel()
+    if not done:
+        raise TimeoutError()
+    if cancel_token.triggered_token is not None:
+        # We've been asked to cancel so we don't care about our future, but we must
+        # consume its exception or else asyncio will emit warnings.
+        [task.exception() for task in done]
```

Comment:
I think this would be better expanded into a two line `for` loop.  Changes this from a *side-effect* based operation to something more explicit.